engine/common/stop/stop_control.go

```diff
@@ -0,0 +1,157 @@
+package stop
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/module/irrecoverable"
+
+	"github.com/onflow/flow-go/engine/common/version"
+)
+
+// ErrNoRegisteredHeightRecorders represents an error indicating that StopControl did not register any execution data height recorders.
+// This error occurs when the StopControl attempts to perform operations that require
+// at least one registered height recorder, but none are found.
+var ErrNoRegisteredHeightRecorders = errors.New("no registered height recorders")
+
+const (
+	DefaultCheckHeightInterval = 30 * time.Second
+)
+
+// StopControl is responsible for managing the stopping behavior of the node
+// when an incompatible block height is encountered.
+type StopControl struct {
+	component.Component
+	cm *component.ComponentManager
+
+	log zerolog.Logger
+
+	// incompatibleBlockHeight is the height of the block that is incompatible with the current node version.
+	incompatibleBlockHeight *atomic.Uint64
+	// updatedVersion is the expected node version to continue working with new blocks.
+	updatedVersion *atomic.String
+
+	registeredHeightRecorders []execution_data.ProcessedHeightRecorder
+}
+
+// NewStopControl creates a new StopControl instance.
+func NewStopControl(
+	log zerolog.Logger,
+	versionControl *version.VersionControl,
+) *StopControl {
+	sc := &StopControl{
+		log: log.With().
+			Str("component", "stop_control").
+			Logger(),
+		incompatibleBlockHeight: atomic.NewUint64(0),
+		updatedVersion:          atomic.NewString(""),
+	}
+
+	sc.cm = component.NewComponentManagerBuilder().
+		AddWorker(sc.loop).
+		Build()
+	sc.Component = sc.cm
+
+	if versionControl != nil {
+		// Subscribe for version updates
+		versionControl.AddVersionUpdatesConsumer(sc.OnVersionUpdate)
+	} else {
+		sc.log.Info().
+			Msg("version control is uninitialized")
+	}
```

Comment:
`versionControl` must be set for stop control to work. You can just skip this an assume it was set. the node will panic when you add the consumer if it's not set